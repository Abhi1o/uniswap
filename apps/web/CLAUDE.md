# CLAUDE.md

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

## Project Overview

This is the Uniswap Web Interface - a React-based decentralized exchange application that's part of the Uniswap Universe monorepo. The web app enables users to swap tokens, provide liquidity, and interact with the Uniswap protocol across multiple blockchain networks.

## Essential Development Commands

### Daily Development

```bash
# Start development server
bun dev

# Run type checking
bun typecheck

# Run linting
bun lint
bun lint:fix

# Run tests
bun run test                    # Run all tests
bun run test:watch             # Watch mode
bun run test:set1              # Components only
bun run test:set2              # Pages and state
bun run test:set3              # Hooks, NFT, utils
bun run test:set4              # Remaining tests

# Run E2E tests
bun playwright:test

# Build for production
bun build:production
```

### Monorepo Commands (from root)

```bash
# Initial setup
bun lfg                    # Full setup with env vars

# Global checks (all packages)
bun g:typecheck
bun g:lint
bun g:test
bun g:build

# Quick checks (changed files only)
bun g:lint:changed
bun g:typecheck:changed
bun g:format:changed
```

## Architecture & Code Organization

### Key Technologies

- **Framework**: React with TypeScript
- **Build**: Vite (primary) with experimental Rolldown, Craco/webpack (legacy)
- **State**: Redux Toolkit, React Query, Jotai
- **Styling**: Styled Components, Tamagui UI framework
- **Web3**: Wagmi, Ethers.js, Web3-React
- **Testing**: Vitest (unit), Playwright (E2E), Storybook (visual)

### Directory Structure

```tree
apps/web/src/
├── components/          # Reusable UI components
├── pages/              # Route-based page components
├── state/              # Redux slices and state logic
├── hooks/              # Custom React hooks
├── connection/         # Web3 wallet connection logic
├── lib/                # External library integrations
├── nft/                # NFT marketplace features
├── utils/              # Utility functions
└── constants/          # App-wide constants
```

### Key Architectural Patterns

1. **Feature-based Organization**: Code is organized by feature/domain rather than by file type
2. **Shared UI Library**: Uses `@uniswap/ui` package from `packages/ui` for consistent components
3. **Strong TypeScript**: Strict typing with comprehensive type definitions
4. **GraphQL Code Generation**: Auto-generated types from GraphQL schemas
5. **Test Colocation**: Unit tests live alongside source files as `.test.ts(x)`

### Testing Strategy

- **Unit Tests**: Use Vitest, focus on logic and hooks
- **Component Tests**: Use React Testing Library with TestID enum
- **E2E Tests**: Playwright tests with `.e2e.test.ts` extension
- **Visual Tests**: Storybook for component documentation and testing

### Important Development Notes

1. **Environment Variables**: Managed via 1Password CLI - run `bun lfg` for setup
2. **Node Version**: Must use Node at the version specified in @.nvmrc
3. **Imports**: Use absolute imports within the app (enforced by ESLint)
4. **TestIDs**: Use the TestID enum instead of string literals for test selectors
5. **GitHub Actions**: External actions must be pinned to commit hashes with version comments

### Common Workflows

**Adding a new feature:**

1. Create feature directory under appropriate section
2. Follow existing patterns for component structure
3. Add Redux slice if state management needed
4. Write tests alongside implementation
5. Update GraphQL queries if needed

**Debugging blockchain interactions:**

```bash
# Start local Ethereum fork
bun anvil:mainnet

# Start local Base fork
bun anvil:base
```

**Working with translations:**

```bash
bun i18n:extract      # Extract new strings
bun i18n:upload       # Upload to Crowdin
bun i18n:download     # Download translations
```

### Code Style Guidelines

- Follow existing patterns in neighboring files
- No unnecessary comments unless explicitly needed
- Use Tamagui components from `@uniswap/ui` when available
- Maintain consistency with Web3 naming conventions
- Always check existing imports before adding new dependencies
