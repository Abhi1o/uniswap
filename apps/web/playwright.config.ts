import { defineConfig, devices } from '@playwright/test'
import dotenv from 'dotenv'
import ms from 'ms'
import path from 'path'

const IS_CI = process.env.CI === 'true'

if (!IS_CI) {
  dotenv.config({ path: path.resolve(__dirname, '.env.local') })
}

// Handle asset files and platform-specific imports for Node.js
// eslint-disable-next-line @typescript-eslint/no-var-requires
const Module = require('module')

// Override module resolution to handle platform-specific files like Vite does
const originalResolveFilename = Module._resolveFilename
Module._resolveFilename = function (request, parent) {
  // For getConfig imports, try .web variant first (mimics Vite behavior)
  // Use precise matching to avoid false positives with modules containing 'getConfig' as substring
  if (request.endsWith('/getConfig') || request.endsWith('\\getConfig') || request === 'getConfig') {
    // Try different .web patterns to match how Vite resolves extensions
    const webVariants = [
      `${request}.web`,
      request.endsWith('.js') ? request.replace(/\.js$/, '.web.js') : `${request}.web.js`,
    ]

    for (const webVariant of webVariants) {
      try {
        return originalResolveFilename.call(this, webVariant, parent)
      } catch {
        // Continue trying other variants
      }
    }
  }

  return originalResolveFilename.call(this, request, parent)
}

const originalLoad = Module._load
Module._load = function (...args: any[]) {
  const [request] = args
  if (request.match(/\.(png|svg|mp4)$/)) {
    return `mock-${path.basename(request)}`
  }
  return originalLoad.apply(this, args)
}

export default defineConfig({
  testDir: './src',
  testMatch: '**/*.e2e.test.ts',
  globalTeardown: './src/playwright/anvil/global-teardown.ts',
  workers: 1, // this is manually configured in the github action depending on type of tests
  fullyParallel: true,
  maxFailures: IS_CI ? 10 : undefined,
  retries: IS_CI ? 3 : 0,
  reporter: IS_CI && process.env.REPORT_TO_SLACK ? [['blob'], ['list']] : 'list',
  timeout: ms('60s'),
  expect: {
    timeout: ms('10s'),
  },
  use: {
    screenshot: 'off',
    video: 'retain-on-failure',
    trace: 'retain-on-failure',
    baseURL: 'http://localhost:3000',
    headless: true,
    extraHTTPHeaders: {
      origin: 'http://localhost:3000',
    },
    launchOptions: {
      args: ['--disable-blink-features=AutomationControlled'],
    },
  },
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] },
    },
  ],
  outputDir: './test-results',
})
