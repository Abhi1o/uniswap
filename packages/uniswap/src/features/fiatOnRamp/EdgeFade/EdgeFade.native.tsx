import { FlexProps, LinearGradient, useSporeColors } from 'ui/src'
import { opacify } from 'ui/src/theme'

export function EdgeFade({ side, width }: { side: 'left' | 'right' } & FlexProps): JSX.Element {
  const colors = useSporeColors()
  return (
    <LinearGradient
      position="absolute"
      top={0}
      bottom={0}
      pointerEvents="none"
      zIndex={1}
      left={side === 'left' ? 0 : undefined}
      right={side === 'right' ? 0 : undefined}
      colors={[opacify(90, colors.surface1.val), opacify(0, colors.surface1.val)]}
      start={{ x: side === 'left' ? 0 : 1, y: 0 }}
      end={{ x: side === 'left' ? 1 : 0, y: 0 }}
      width={width}
    />
  )
}
