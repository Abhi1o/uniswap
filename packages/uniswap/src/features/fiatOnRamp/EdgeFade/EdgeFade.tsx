import { FlexProps } from 'ui/src'
import { PlatformSplitStubError } from 'utilities/src/errors'

export function EdgeFade(_props: { side: 'left' | 'right' } & FlexProps): JSX.Element {
  throw new PlatformSplitStubError('EdgeFade')
}
