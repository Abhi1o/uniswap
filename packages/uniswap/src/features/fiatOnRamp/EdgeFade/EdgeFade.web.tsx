import { Flex, FlexProps, useSporeColors } from 'ui/src'
import { opacify } from 'ui/src/theme'

export function EdgeFade({ side, ...rest }: { side: 'left' | 'right' } & FlexProps): JSX.Element {
  const colors = useSporeColors()
  return (
    <Flex
      position="absolute"
      top={0}
      bottom={0}
      pointerEvents="none"
      zIndex={1}
      left={side === 'left' ? 0 : undefined}
      right={side === 'right' ? 0 : undefined}
      style={{
        background: `linear-gradient(${side === 'left' ? '90deg' : '270deg'}, ${opacify(90, colors.surface1.val)} 0%, ${opacify(0, colors.surface1.val)} 100%)`,
      }}
      {...rest}
    />
  )
}
