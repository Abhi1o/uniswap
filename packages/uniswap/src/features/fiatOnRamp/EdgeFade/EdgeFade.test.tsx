import { EdgeFade as MobileEdgeFade } from 'uniswap/src/features/fiatOnRamp/EdgeFade/EdgeFade.native'
import { EdgeFade as WebEdgeFade } from 'uniswap/src/features/fiatOnRamp/EdgeFade/EdgeFade.web'
import { render } from 'uniswap/src/test/test-utils'

describe('EdgeFade', () => {
  const leftProps = {
    side: 'left' as const,
  }
  const rightProps = {
    side: 'right' as const,
  }

  describe('Mobile Implementation', () => {
    it('renders mobile implementation correctly with left side', () => {
      const tree = render(<MobileEdgeFade {...leftProps} />)
      expect(tree).toMatchSnapshot()
    })

    it('renders mobile implementation correctly with right side', () => {
      const propsWithRightSide = {
        ...rightProps,
      }
      const tree = render(<MobileEdgeFade {...propsWithRightSide} />)
      expect(tree).toMatchSnapshot()
    })

    it('renders with additional flex props', () => {
      const propsWithFlexProps = {
        ...leftProps,
        testID: 'mobile-edge-fade',
        width: 24,
      }
      const tree = render(<MobileEdgeFade {...propsWithFlexProps} />)
      expect(tree).toMatchSnapshot()
    })
  })

  describe('Web Implementation', () => {
    it('renders web implementation correctly with left side', () => {
      const tree = render(<WebEdgeFade {...leftProps} />)
      expect(tree).toMatchSnapshot()
    })

    it('renders web implementation correctly with right side', () => {
      const propsWithRightSide = {
        ...rightProps,
      }
      const tree = render(<WebEdgeFade {...propsWithRightSide} />)
      expect(tree).toMatchSnapshot()
    })

    it('renders with additional flex props', () => {
      const propsWithFlexProps = {
        ...leftProps,
        testID: 'web-edge-fade',
        width: 24,
      }
      const tree = render(<WebEdgeFade {...propsWithFlexProps} />)
      expect(tree).toMatchSnapshot()
    })
  })
})
