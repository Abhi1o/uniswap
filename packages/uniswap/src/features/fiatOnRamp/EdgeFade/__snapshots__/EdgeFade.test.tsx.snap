// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EdgeFade Mobile Implementation renders mobile implementation correctly with left side 1`] = `
<View
  pointerEvents="none"
  style={
    {
      "bottom": 0,
      "flexDirection": "column",
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "zIndex": 1,
    }
  }
>
  ExpoLinearGradient
</View>
`;

exports[`EdgeFade Mobile Implementation renders mobile implementation correctly with right side 1`] = `
<View
  pointerEvents="none"
  style={
    {
      "bottom": 0,
      "flexDirection": "column",
      "overflow": "hidden",
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 1,
    }
  }
>
  ExpoLinearGradient
</View>
`;

exports[`EdgeFade Mobile Implementation renders with additional flex props 1`] = `
<View
  pointerEvents="none"
  style={
    {
      "bottom": 0,
      "flexDirection": "column",
      "left": 0,
      "overflow": "hidden",
      "position": "absolute",
      "top": 0,
      "width": 24,
      "zIndex": 1,
    }
  }
>
  ExpoLinearGradient
</View>
`;

exports[`EdgeFade Web Implementation renders web implementation correctly with left side 1`] = `
<View
  pointerEvents="none"
  style={
    {
      "background": "linear-gradient(90deg, #131313e6 0%, #13131300 100%)",
      "bottom": 0,
      "flexDirection": "column",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "zIndex": 1,
    }
  }
/>
`;

exports[`EdgeFade Web Implementation renders web implementation correctly with right side 1`] = `
<View
  pointerEvents="none"
  style={
    {
      "background": "linear-gradient(270deg, #131313e6 0%, #13131300 100%)",
      "bottom": 0,
      "flexDirection": "column",
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 1,
    }
  }
/>
`;

exports[`EdgeFade Web Implementation renders with additional flex props 1`] = `
<View
  pointerEvents="none"
  style={
    {
      "background": "linear-gradient(90deg, #131313e6 0%, #13131300 100%)",
      "bottom": 0,
      "flexDirection": "column",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": 24,
      "zIndex": 1,
    }
  }
  testID="web-edge-fade"
/>
`;
